### PR TITLE
#21838: Fix ByteStringParser backpressure

### PR DESCRIPTION
**Pull request by drewhk**
and convert it to simple push-pull

### CODE REVIEW

**Comment by drewhk** [comment]
fixes #21838

File: akka-stream-tests/src/test/scala/akka/stream/io/ByteStringParserSpec.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package akka.stream.io
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/io/ByteStringParserSpec.scala]: no header? 😢

File: akka-stream-tests/src/test/scala/akka/stream/io/ByteStringParserSpec.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+      startWith(step)
+
+    }
+
+  }
+
+  "ByteStringParser" must {
+
+    "respect backpressure" in {
```
Comment (jrudolph) [inline akka-stream-tests/src/test/scala/akka/stream/io/ByteStringParserSpec.scala]: Maybe add a comment that explains how that stream setup tests that backpressure works correctly.

File: akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala [hunk 3]
```diff
@@ -22,54 +22,84 @@ private[akka] abstract class ByteStringParser[T] extends GraphStage[FlowShape[By
...
+    override def onPull(): Unit = doParse()
 
     def onPush(): Unit = {
-      pullOnParserRequest = false
+      // Buffer management before we call doParse():
+      //  - append new bytes
+      //  - compact buffer if necessary
       buffer ++= grab(bytesIn)
+      untilCompact -= 1
+      if (untilCompact == 0) {
```
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala]: Do you have an idea how much of an effect this compacting would have? It seems rather arbitrary to compact on all 16 incoming elements. It's rather more important that outgoing ByteStrings are compacted (but this stage can't do anything to enforce that).

**Review by jrudolph — APPROVED** [review]
LGTM. Good to see that this is hardly more complex than before.

File: akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala]: The issue is that ever-ovlerlapping slices of the buffer can end up collecting an ever-growing tree. This has nothing to do with outgoing ByteStrings.

File: akka-stream-tests/src/test/scala/akka/stream/io/ByteStringParserSpec.scala (hunk 2 above)
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/io/ByteStringParserSpec.scala]: Ok, the idea is to emit more than one downstream element per upstream chunk.

File: akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala]: Basically the threshold is a pruning logic that amortizes copy cost but keeps ByteString nesting levels at a bounded level.

File: akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala [hunk 4]
```diff
@@ -78,6 +111,8 @@ private[akka] abstract class ByteStringParser[T] extends GraphStage[FlowShape[By
  */
 private[akka] object ByteStringParser {
 
+  val CompactionThreshold = 16
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala]: Science behind magic number?

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala]: Contacted my astrology advisor :D I will be honest, I don't know what is the ideal number here. It should be a finite number and not too small to reduce the copying.

File: akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala [hunk 5]
```diff
@@ -22,54 +22,87 @@ private[akka] abstract class ByteStringParser[T] extends GraphStage[FlowShape[By
   override def initialAttributes = Attributes.name("ByteStringParser")
   final override val shape = FlowShape(bytesIn, objOut)
 
-  class ParsingLogic extends GraphStageLogic(shape) with InHandler {
-    var pullOnParserRequest = false
-    override def preStart(): Unit = pull(bytesIn)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala]: not pulling in preStart is by design, I assume

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala (hunk 4 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala]: I propose an alternative solution as this is a common issue: https://github.com/akka/akka/issues/21865

I will fix it separately as the fix in this PR is important.

File: akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala (hunk 5 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala]: Yes, no need for detached operation.

**Pull request merged**